lab 8: перенос обучения для eurosat

цели:
- загрузить набор данных eurosat
- выбрать предобученную сеть классификации изображений
- заморозить базовые слои, добавить новые слои и обучить их на eurosat
- показать качество на проверочной выборке

In [1]:
import torch
from torch import nn
from torch.utils.data import DataLoader, random_split, Subset
import torchvision
from torchvision import transforms
from torchvision.datasets import EuroSAT
from torchvision.datasets.utils import download_and_extract_archive
from torchvision.models import resnet18, ResNet18_Weights
from PIL import Image
from pathlib import Path
import shutil
import time

torch.manual_seed(42)
ustroystvo = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
ustroystvo

device(type='cpu')

подготовка данных

In [2]:
katalog_dannykh = Path('data')
bazovye_preobrazovaniya = transforms.Compose([
    transforms.Resize((224, 224)),
    transforms.ToTensor(),
    transforms.Normalize(mean=[0.485, 0.456, 0.406], std=[0.229, 0.224, 0.225]),
])

try:
    polnyy_nabor = EuroSAT(root=katalog_dannykh, download=True, transform=bazovye_preobrazovaniya)
    chislo_klassov = len(polnyy_nabor.classes)
except Exception as oshibka_zagruzki:
    print('скачивание через torchvision не удалось, пробую http-зеркало:', oshibka_zagruzki)
    baza_http = katalog_dannykh / 'eurosat'
    if (katalog_dannykh / 'EuroSAT').exists():
        shutil.rmtree(katalog_dannykh / 'EuroSAT')
    baza_http.mkdir(parents=True, exist_ok=True)
    rezervnaya_ssylka = 'http://madm.dfki.de/files/sentinel/EuroSAT.zip'
    download_and_extract_archive(
        rezervnaya_ssylka,
        download_root=baza_http,
        extract_root=baza_http,
        filename='EuroSAT_http.zip',
    )
    polnyy_nabor = EuroSAT(root=katalog_dannykh, download=False, transform=bazovye_preobrazovaniya)
    chislo_klassov = len(polnyy_nabor.classes)

razmer_podborki = 600
esli_nuzhno = torch.randperm(len(polnyy_nabor))[:razmer_podborki]
podborka = Subset(polnyy_nabor, esli_nuzhno)

razmer_obucheniya = int(0.7 * len(podborka))
razmer_proverki = int(0.15 * len(podborka))
razmer_test = len(podborka) - razmer_obucheniya - razmer_proverki

obucheniye_nabor, proverka_nabor, test_nabor = random_split(
    podborka,
    [razmer_obucheniya, razmer_proverki, razmer_test],
    generator=torch.Generator().manual_seed(42),
)

obucheniye_zagruzchik = DataLoader(obucheniye_nabor, batch_size=32, shuffle=True, num_workers=0)
proverka_zagruzchik = DataLoader(proverka_nabor, batch_size=32, shuffle=False, num_workers=0)
test_zagruzchik = DataLoader(test_nabor, batch_size=32, shuffle=False, num_workers=0)

razmer_obucheniya, razmer_proverki, razmer_test

(420, 90, 90)

модель

In [3]:
bazovaya_set = resnet18(weights=ResNet18_Weights.IMAGENET1K_V1)
for param in bazovaya_set.parameters():
    param.requires_grad = False

izvlecheniye_priznakov = list(bazovaya_set.children())[:-1]

novyy_klassifikator = nn.Sequential(
    nn.Flatten(),
    nn.Linear(bazovaya_set.fc.in_features, 256),
    nn.ReLU(),
    nn.Dropout(0.3),
    nn.Linear(256, chislo_klassov),
)

model = nn.Sequential(*izvlecheniye_priznakov, novyy_klassifikator)
model.to(ustroystvo)

kriteriy = nn.CrossEntropyLoss()
optimizator = torch.optim.Adam(novyy_klassifikator.parameters(), lr=1e-3)
model

Sequential(
  (0): Conv2d(3, 64, kernel_size=(7, 7), stride=(2, 2), padding=(3, 3), bias=False)
  (1): BatchNorm2d(64, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
  (2): ReLU(inplace=True)
  (3): MaxPool2d(kernel_size=3, stride=2, padding=1, dilation=1, ceil_mode=False)
  (4): Sequential(
    (0): BasicBlock(
      (conv1): Conv2d(64, 64, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1), bias=False)
      (bn1): BatchNorm2d(64, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
      (relu): ReLU(inplace=True)
      (conv2): Conv2d(64, 64, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1), bias=False)
      (bn2): BatchNorm2d(64, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
    )
    (1): BasicBlock(
      (conv1): Conv2d(64, 64, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1), bias=False)
      (bn1): BatchNorm2d(64, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
      (relu): ReLU(inplace=True)
      (conv2): Con

обучение

In [4]:
def epokha_obucheniya():
    model.train()
    obshchaya_poterja = 0.0
    pravilno = 0
    vsego = 0
    for izobrazheniya, metki in obucheniye_zagruzchik:
        izobrazheniya, metki = izobrazheniya.to(ustroystvo), metki.to(ustroystvo)
        optimizator.zero_grad()
        vyrabotka = model(izobrazheniya)
        poterja = kriteriy(vyrabotka, metki)
        poterja.backward()
        optimizator.step()
        obshchaya_poterja += poterja.item() * izobrazheniya.size(0)
        _, prognoz = torch.max(vyrabotka, 1)
        vsego += metki.size(0)
        pravilno += (prognoz == metki).sum().item()
    return obshchaya_poterja / vsego, pravilno / vsego

def ocenit(zagruzchik):
    model.eval()
    obshchaya_poterja = 0.0
    pravilno = 0
    vsego = 0
    with torch.no_grad():
        for izobrazheniya, metki in zagruzchik:
            izobrazheniya, metki = izobrazheniya.to(ustroystvo), metki.to(ustroystvo)
            vyrabotka = model(izobrazheniya)
            poterja = kriteriy(vyrabotka, metki)
            obshchaya_poterja += poterja.item() * izobrazheniya.size(0)
            _, prognoz = torch.max(vyrabotka, 1)
            vsego += metki.size(0)
            pravilno += (prognoz == metki).sum().item()
    return obshchaya_poterja / vsego, pravilno / vsego

epokhi = 3
itogi = []
dlya_epokhi = []
start = time.time()
for nomer in range(1, epokhi + 1):
    poterja_obuch, tochnost_obuch = epokha_obucheniya()
    poterja_prov, tochnost_prov = ocenit(proverka_zagruzchik)
    dlya_epokhi.append((nomer, poterja_obuch, tochnost_obuch, poterja_prov, tochnost_prov))
    itogi.append({
        'epokha': nomer,
        'poterja_obucheniye': round(poterja_obuch, 4),
        'tochnost_obucheniye': round(tochnost_obuch, 4),
        'poterja_proverka': round(poterja_prov, 4),
        'tochnost_proverka': round(tochnost_prov, 4),
    })
end = time.time()
print('время обучения, с:', round(end - start, 2))
itogi

время обучения, с: 99.95


[{'epokha': 1,
  'poterja_obucheniye': 2.0016,
  'tochnost_obucheniye': 0.3333,
  'poterja_proverka': 1.683,
  'tochnost_proverka': 0.5556},
 {'epokha': 2,
  'poterja_obucheniye': 1.3336,
  'tochnost_obucheniye': 0.65,
  'poterja_proverka': 1.1828,
  'tochnost_proverka': 0.6444},
 {'epokha': 3,
  'poterja_obucheniye': 1.0353,
  'tochnost_obucheniye': 0.681,
  'poterja_proverka': 0.886,
  'tochnost_proverka': 0.7889}]

проверка на тесте

In [5]:
poterja_test, tochnost_test = ocenit(test_zagruzchik)
{'poterja_test': round(poterja_test, 4), 'tochnost_test': round(tochnost_test, 4)}

{'poterja_test': 0.8196, 'tochnost_test': 0.7222}

выводы

выводы по эксперименту:- за три эпохи на подвыборке 600 снимков потери на обучении упали с 2.00 до 1.04, точность выросла до 0.6810, а на проверке потеря снизилась до 0.8860 при точности 0.7889.- итоговая проверка на тесте дала poterja_test=0.8196 и tochnost_test=0.7222, что подтверждает усвоение классов, но оставляет запас для улучшений.- увеличение числа эпох и данных или разморозка части бэкбона должны поднять точность к значениям 0.85+ на тесте.

статус: лабораторная работа завершена, тетрадь готова к проверке.